File: UCF HSPT Documents/2007/Solutions/marquee.cpp
/*
UCF's 2007 High School Programming Contest
Marquee
Solution Prepared by: Nick Beato

To solve this problem, we will simulate the marquee sign in a string.  The
cases we need to watch out for are:

1) When the number of characters that a sign may display is equal to the message length.
In this case, we simply output the message.

2) When the number of characters that a sign may display is larger than the message length.
In this case, we need to append spaces to the end of the input message, reducing it to the
previous case.

3) When the message will not fit on the sign.  In this case, we simulate the marquee.  Since
there is a space separating the end of the message from the beginning during a cycle, we can
directly append this to the message.  Notice that the number of lines in the output is then
equal to the message length, so we known exactly how many times to cycle the message.

With that, let's look at the solution.

*/

#include <iostream>
#include <string>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
	ifstream in("marquee.in");
        string tmp;

	// n in "the input"
	int numberOfSigns;
        getline(in, tmp);
        numberOfSigns = atoi(tmp.c_str());

	// process all n signs, output each sign as it is processed
	for(int sign = 1; sign <= numberOfSigns; sign++) {
		// the sign's message
		string message;

		// the number of characters this sign can display
		int capacity;

		// read the sign information.
		// cin.get() is invoked to remove the '\n' character
		// from the line with the previous integer.
		getline(in, message);
                getline(in, tmp);
                capacity = atoi(tmp.c_str());

		// output the header
		cout << "Sign #" << sign << ":" << endl;

		// case 1 and case 2
		if(capacity >= message.length()) {

			// reduces case 2 to case 1 by adding spaces
			while(message.length() < capacity)
				message += " ";

			cout << "[" << message << "]" << endl;
		}

		// case 3
		else {
			// add the extra space during the cycle
			message += " ";

			// simulate the marquee by outputing the sign
			// then moving the first character to the end of the string
			for(int i = 0; i < message.length(); i++) {
				cout << "[" << message.substr(0, capacity) << "]" << endl;
				message = message.substr(1) + message[0];
			}
		}

		cout << endl;
	}

	in.close();
	return 0;
}

File: UCF HSPT Team/src/y2007/marquee.java
/* *****************************************
 * Problem: QUEE MAR
 * 
 * This problem is pretty simple: just rotate the string for one full
 * rotation!
 *******************************************/
import java.util.*;
import java.io.*;


public class marquee {
	public static void main(String[] args) throws IOException
	{
		// Open the input file for reading
		Scanner fin = new Scanner(new File("marquee.in"));
		int signs = fin.nextInt();
		
		// Process each sign
		for(int sign = 1; sign <= signs; sign++)
		{
			fin.nextLine(); // Consume the line separator
			String s = fin.nextLine(); // Read the string to rotate
			int m = fin.nextInt(); // Read the length of the marquee
			
			System.out.println("Sign #"+sign+":"); // Print the sign header
			
			// If the message is short enough to fit on the sign, just print
			// it out
			if(m >= s.length())
			{
			        // If the sign is bigger than the message, we
				// need to add spaces to fill the sign
				while (m > s.length())
					s += " ";

				System.out.println("["+s+"]");
				System.out.println();
				continue;
			}
			
			s = s+" ";
			// The number of lines of output is equal to the number of 
			// characters in the string we're rotating
			for(int offset = 0; offset < s.length(); offset++)
			{
				System.out.print("[");
				// Print out enough characters to fill the marquee, beginning
				// with character at position 'offset' and wrapping around if
				// need be.
				for(int i = 0; i < m; i++)
				{
					System.out.print(s.charAt((i+offset)%s.length()));
				}
				System.out.println("]"); // Wrap up the string
			}
			System.out.println();
		}
	}
}
